onnxruntime/core/providers/cuda/nn/conv.cc

```diff
@@ -88,15 +114,46 @@ Status Conv<T>::ComputeInternal(OpKernelContext* context) const {
       }
 
       std::vector<int64_t> y_dims;
+      y_dims.reserve(2 + rank);  // rank indicates number of feature dimensions - so add 2 to account for 'N' and 'C'
       y_dims.insert(y_dims.begin(), {N, M});
-      ORT_RETURN_IF_ERROR(conv_attrs_.InferOutputShape(x_shape.Slice(2), kernel_shape,
-                                                       strides, dilations, pads, y_dims, true));
+
+      std::vector<int64_t> y_dims_with_adjusted_pads;
+      y_dims_with_adjusted_pads.reserve(2 + rank);  // rank indicates number of feature dimensions - so add 2 to account for 'N' and 'C'
+      y_dims_with_adjusted_pads.insert(y_dims_with_adjusted_pads.begin(), {N, M});
+
+      bool post_slicing_required = false;
+      std::vector<int64_t> slice_starts;
+      slice_starts.reserve(rank);
+
+      std::vector<int64_t> slice_ends;
+      slice_ends.reserve(rank);
+
+      std::vector<int64_t> slice_axes;
+      slice_axes.reserve(rank);
+
+      ORT_RETURN_IF_ERROR(conv_attrs_.InferOutputShapeWithAdjustedPads(x_shape.Slice(2), kernel_shape,
+                                                                       strides, dilations, pads, y_dims, y_dims_with_adjusted_pads,
+                                                                       post_slicing_required, slice_starts, slice_ends, slice_axes));
+      ORT_ENFORCE(y_dims.size() == y_dims_with_adjusted_pads.size());
       s_.y_dims = y_dims;
-      Tensor* Y = context->Output(0, TensorShape(s_.y_dims));
-      y_data = reinterpret_cast<CudaT*>(Y->template MutableData<T>());
+      s_.y_dims_with_adjusted_pads = y_dims_with_adjusted_pads;
+      s_.post_slicing_required = post_slicing_required;
+      s_.slice_starts = slice_starts;
+      s_.slice_ends = slice_ends;
+      s_.slice_axes = slice_axes;
+
+      Y = context->Output(0, TensorShape(s_.y_dims));
+      if (!post_slicing_required) {
+        // No post slicing needed. Fill the output tensor's buffer directly.
+        y_data = reinterpret_cast<CudaT*>(Y->template MutableData<T>());
+      } else {
+        // Post slicing needed. Create and fill in the Conv results in an intermediate buffer.
+        y_data = reinterpret_cast<CudaT*>(s_.AllocMemoryForCuDnnConvResults(allocator,
+                                                                            TensorShape(y_dims_with_adjusted_pads).Size() * element_size));
+      }
```

Comment:
Not sure this is safe. What guarantee's that y_data will be unused when the lock is released? My understanding was that the CUDA operations involving this may execute asynchronously after ComputeInternal returns.

Reply:
I see. I half suspected that this was the case myself. This is probably why `workspace` is allocated on every Run() as well. Otherwise, we may as well allocate that once per kernel and re-use that given that pretty much the entire Run() is being done after aquiring a lock. 

I ll follow the same approach as `workspace` for `y_data`. I guess allocating once per Run() is not the end of the world when used in conjunction with an arena which greatly reduces the overhead of "allocating" memory. Is that okay ?

Reply:
The temporary buffer (y_data) is allocated on every Run(). No more allocating and caching.